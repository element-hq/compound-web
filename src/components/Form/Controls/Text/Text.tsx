/*
Copyright 2023 The Matrix.org Foundation C.I.C.
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { forwardRef, ComponentProps, ComponentRef } from "react";
import { Control } from "@radix-ui/react-form";

import styles from "./Text.module.css";
import classNames from "classnames";

type TextProps = {
  /**
   * The CSS class name.
   */
  className?: string;

  /**
   * Enable contextual alternate ligatures on input text
   * For example on an in-place editing field
   * https://github.com/rsms/inter/issues/222
   * https://github.com/rsms/inter/blob/master/src/features/calt.fea
   */
  enableLigatures?: boolean;
} & ComponentProps<"input">;

/**
 * A styled text input, for standalone use.
 */
export const TextInput = forwardRef<HTMLInputElement, TextProps>(
  function TextInput(props, ref) {
    const classes = classNames(
      styles.control,
      props.className,
      props.enableLigatures && styles["enable-ligatures"],
    );
    return <input ref={ref} {...props} className={classes} />;
  },
);

/**
 * A styled text input wrapped in a `Control` component, for use in Radix forms.
 */
export const TextControl = forwardRef<
  ComponentRef<typeof TextInput>,
  ComponentProps<typeof TextInput>
>(function TextControl(props, ref) {
  return (
    <Control asChild>
      <TextInput ref={ref} {...props} />
    </Control>
  );
});
