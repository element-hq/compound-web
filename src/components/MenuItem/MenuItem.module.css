/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.item {
  display: grid;
  grid-template: "icon label ." auto "empty1 label empty2" auto / auto auto minmax(
      /* Reserve space for the chevron so that the layout doesn't shift on
      hover */
        var(--cpd-space-4x),
      1fr
    );
  align-items: center;
  justify-items: end;
  padding-block: var(--cpd-space-2x);
  padding-inline: var(--cpd-space-4x);
  box-sizing: border-box;
  inline-size: 100%;
  min-inline-size: 200px;
  color: var(--cpd-color-text-secondary);
  background: var(--cpd-color-bg-action-secondary-rest);
}

.item.interactive {
  cursor: pointer;
}

.item.no-label {
  grid-template: "icon ." auto / auto 1fr;
}

/* Items with interactive controls like toggles are larger by default */
label.item {
  min-inline-size: 246px;
}

.icon {
  grid-area: icon;
  margin-inline-end: var(--cpd-space-3x);
}

.item.no-label .icon {
  margin-inline-end: var(--cpd-space-4x);
}

.label {
  grid-area: label;
  margin-inline-end: var(--cpd-space-4x);
  text-align: start;
  word-break: break-word;
}

.nav-hint {
  /* Hidden until the item is hovered over */
  display: none;
  flex-shrink: 0;
  margin-inline-end: calc(-1 * var(--cpd-space-2x));
}

button.item {
  appearance: none;
  border: none;
}

.item[data-kind="primary"] > .label {
  color: var(--cpd-color-text-primary);
}

.item[data-kind="primary"] > .icon {
  color: var(--cpd-color-icon-primary);
}

.item[data-kind="primary"] > .nav-hint {
  color: var(--cpd-color-icon-tertiary);
}

.item[data-kind="critical"] > .label {
  color: var(--cpd-color-text-critical-primary);
}

.item[data-kind="critical"] > .icon,
.item[data-kind="critical"] > .nav-hint {
  color: var(--cpd-color-icon-critical-primary);
}

@media (hover) {
  .item.interactive[data-kind="primary"]:hover {
    background: var(--cpd-color-bg-action-secondary-hovered);
  }

  .item.interactive[data-kind="critical"]:hover {
    background: var(--cpd-color-bg-critical-subtle);
  }

  /* Replace the children with the navigation hint on hover */
  .item.interactive:hover > .nav-hint {
    display: initial;
  }

  .item.interactive:hover > .nav-hint ~ * {
    display: none;
  }
}

.item.interactive[data-kind="primary"]:active {
  background: var(--cpd-color-bg-action-secondary-pressed);
}

.item.interactive[data-kind="critical"]:active {
  background: var(--cpd-color-bg-critical-subtle);
}

.item[data-kind].disabled {
  pointer-events: none;
}

.item[data-kind].disabled > .label,
.item[data-kind].disabled > .icon,
.item[data-kind].disabled > .nav-hint {
  color: var(--cpd-color-text-disabled);
}
