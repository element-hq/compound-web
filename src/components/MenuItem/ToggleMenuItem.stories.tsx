/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { Meta, StoryFn } from "@storybook/react";
import ExtensionsIcon from "@vector-im/compound-design-tokens/icons/extensions.svg";
import ChatIcon from "@vector-im/compound-design-tokens/icons/chat.svg";

import { ToggleMenuItem as ToggleMenuItemComponent } from "./ToggleMenuItem";

export default {
  title: "ToggleMenuItem",
  component: ToggleMenuItemComponent,
  tags: ["autodocs"],
  argTypes: {},
  args: {},
} as Meta<typeof ToggleMenuItemComponent>;

const Template: StoryFn<typeof ToggleMenuItemComponent> = (args) => (
  <div style={{ width: 300 }}>
    <ToggleMenuItemComponent {...args} Icon={ChatIcon} label="First item" />
    <ToggleMenuItemComponent
      {...args}
      Icon={ExtensionsIcon}
      label="Second item with a name that's quite long"
    />
  </div>
);

export const Primary = Template.bind({});
Primary.args = {};
