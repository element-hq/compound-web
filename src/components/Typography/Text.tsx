/*
Copyright 2025 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import { Typography } from "./Typography";

type TypographyProps = React.ComponentProps<typeof Typography>;

/**
 * A text component. Underlying HTML element can be changed using the `as`
 * property. Will default to be a paragraph.
 */
export const Text: React.FC<
  Omit<TypographyProps, "type"> & {
    // xl is not a valid text size
    size?: Exclude<TypographyProps["size"], "xl">;
  }
> = ({ as = "p", children, ...props }) => {
  return (
    <Typography as={as} type="body" {...props}>
      {children}
    </Typography>
  );
};
