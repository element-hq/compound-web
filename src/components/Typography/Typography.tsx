/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { PropsWithChildren } from "react";
import classNames from "classnames";

import styles from "./Typography.module.css";

type TypographyProps<C extends React.ElementType> = {
  /**
   * The HTML tag.
   * @default "p"
   */
  as?: C;
  /**
   * The type of content.
   */
  type?: "body" | "heading";
  /**
   * The font weight.
   */
  weight?: "regular" | "semibold" | "medium" | "bold";
  /**
   * The t-shirt size of the content.
   */
  size?: "xs" | "sm" | "md" | "lg" | "xl";
  /**
   * The CSS class name.
   */
  className?: string;
} & React.ComponentPropsWithoutRef<C>;

export const Typography = <C extends React.ElementType = "p">({
  as,
  children,
  type = "body",
  weight = "regular",
  size = "md",
  className,
  ...restProps
}: PropsWithChildren<TypographyProps<C>>): React.ReactElement => {
  const Component = as || "p";

  return (
    <Component
      {...restProps}
      className={classNames(
        styles.typography,
        styles[`font-${type}-${size}-${weight}`],
        className,
      )}
    >
      {children}
    </Component>
  );
};
