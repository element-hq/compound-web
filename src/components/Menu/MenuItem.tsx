/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import classnames from "classnames";
import React, {
  ComponentPropsWithoutRef,
  ComponentType,
  ElementType,
  isValidElement,
  ReactElement,
  SVGAttributes,
  useCallback,
  useContext,
  MouseEventHandler,
} from "react";
import styles from "./MenuItem.module.css";
import { Text } from "../Typography/Text";
import ChevronRightIcon from "@vector-im/compound-design-tokens/assets/web/icons/chevron-right";
import { MenuContext } from "./MenuContext";
import { Slot } from "@radix-ui/react-slot";

type MenuItemElement = "button" | "a" | "div";

type Props<C extends MenuItemElement> = {
  /**
   * The element type of this menu item.
   * @default button
   */
  as?: C;
  /**
   * The CSS class name.
   */
  className?: string;
  /**
   * The icon to show on this menu item.
   * When `Icon` is a ReactElement, it should spread the props
   */
  Icon?: ComponentType<SVGAttributes<SVGElement>> | ReactElement;
  /**
   * The label to show on this menu item.
   */
  // This prop is required because it's rare to not want a label
  label: string | null;
  /**
   * Additional properties to pass to the Text label component.
   */
  labelProps?: ComponentPropsWithoutRef<typeof Text>;
  /**
   * Event callback for when the item is selected via mouse, touch, or keyboard.
   * Calling event.preventDefault in this handler will prevent the menu from
   * being dismissed.
   */
  // This prop is required because it's rare to not want a selection handler
  onSelect: ((e: Event) => void) | null;
  /**
   * Event callback for when the item is clicked.
   * @param e
   */
  onClick?: MouseEventHandler<HTMLElementTagNameMap[C]>;
  /**
   * The color variant of the menu item.
   * @default primary
   */
  kind?: "primary" | "critical";
  disabled?: boolean;
  /**
   * Whether to hide the chevron navigation hint.
   */
  hideChevron?: boolean;
} & Omit<ComponentPropsWithoutRef<C>, "onSelect" | "onClick">;

/**
 * An item within a menu, acting either as a navigation button, or simply a
 * container for other interactive elements.
 * Must be used within a compound Menu or other `menu` or `menubar` aria role subtree.
 */
export const MenuItem = <C extends MenuItemElement = "button">({
  as,
  className,
  Icon,
  label,
  labelProps,
  onSelect,
  kind = "primary",
  children,
  onClick: onClickProp,
  disabled,
  hideChevron,
  ...props
}: Props<C>): React.ReactElement => {
  const Component = as ?? ("button" as ElementType);
  const context = useContext(MenuContext);

  const onClick = useCallback(
    (e: Parameters<Exclude<typeof onClickProp, undefined>>[0]) => {
      (onClickProp as ((e_: typeof e) => void) | undefined)?.(e);
      // If there is no wrapper component to automatically handle onSelect, we
      // need to handle it manually, dismissing the menu as the default action
      if (onSelect !== null && context?.MenuItemWrapper == null) {
        const selectEvent = new CustomEvent("menu.itemSelect", {
          bubbles: true,
          cancelable: true,
        });
        onSelect(selectEvent);
        if (!selectEvent.defaultPrevented) context?.onOpenChange(false);
      }
    },
    [context, onSelect],
  );

  const iconIsReactElement = isValidElement(Icon);
  const componentIcon = Icon as ReactElement;
  const SvgIcon = Icon as ComponentType<SVGAttributes<SVGElement>>;

  const content = (
    <Component
      role="menuitem"
      {...props}
      className={classnames(className, styles.item, {
        [styles.interactive]: onSelect !== null,
        [styles["no-label"]]: label === null,
        [styles["no-icon"]]: !Icon,
        [styles["disabled"]]: disabled,
      })}
      data-kind={kind}
      onClick={onClick}
      disabled={Component === "button" ? disabled : undefined}
      aria-disabled={Component === "button" ? undefined : disabled}
    >
      {Icon &&
        (iconIsReactElement ? (
          <Slot className={styles.icon}>{componentIcon}</Slot>
        ) : (
          <SvgIcon
            width={24}
            height={24}
            className={styles.icon}
            aria-hidden={true}
          />
        ))}

      {label !== null && (
        <Text
          className={styles.label}
          size="md"
          weight="medium"
          as="span"
          {...labelProps}
        >
          {label}
        </Text>
      )}
      {/* We use CSS to swap between this navigation hint and the provided
      children on hover - see the styles module. */}
      {!hideChevron && (Component === "button" || Component === "a") && (
        <ChevronRightIcon
          width={8}
          height={24}
          className={styles["nav-hint"]}
          aria-hidden={true}
          /* The SVG is a small icon in a large canvas. It probably ought to be
             cropped, but we can adjust the viewBox here to chop off the horizontal
             space to get it closer to the right hand edge.
           */
          viewBox="8 0 8 24"
        />
      )}
      {children}
    </Component>
  );

  return context?.MenuItemWrapper == null || onSelect === null ? (
    content
  ) : (
    <context.MenuItemWrapper onSelect={onSelect}>
      {content}
    </context.MenuItemWrapper>
  );
};
