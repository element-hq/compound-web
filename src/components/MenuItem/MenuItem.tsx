/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classnames from "classnames";
import React, {
  ComponentPropsWithoutRef,
  ComponentType,
  ElementType,
  SVGAttributes,
} from "react";
import styles from "./MenuItem.module.css";
import { Text } from "../Typography/Text";
import ChevronRightIcon from "@vector-im/compound-design-tokens/icons/chevron-right.svg";

type MenuItemElement = "button" | "label" | "a" | "div";

type Props<C extends MenuItemElement> = {
  /**
   * The element type of this menu item.
   * @default button
   */
  as?: C;
  /**
   * The CSS class name.
   */
  className?: string;
  /**
   * The icon to show on this menu item.
   */
  Icon: ComponentType<SVGAttributes<SVGElement>>;
  /**
   * The label to show on this menu item.
   */
  // This prop is required because it's rare to not want a label
  label: string | undefined;
  /**
   * The color variant of the menu item.
   * @default primary
   */
  kind?: "primary" | "critical";

  disabled?: boolean;
} & ComponentPropsWithoutRef<C>;

/**
 * An item within a menu, acting either as a navigation button, or simply a
 * container for other interactive elements.
 */
export const MenuItem = <C extends MenuItemElement = "button">({
  as,
  className,
  Icon,
  label,
  kind = "primary",
  children,
  disabled,
  ...props
}: Props<C>): JSX.Element => {
  const Component = as ?? ("button" as ElementType);

  return (
    <Component
      role="menuitem"
      {...props}
      className={classnames(className, styles.item, {
        [styles.interactive]: as !== "div",
        [styles["no-label"]]: label === undefined,
        [styles["disabled"]]: disabled,
      })}
      data-kind={kind}
      disabled={disabled}
    >
      <Icon width={24} height={24} className={styles.icon} aria-hidden={true} />
      {label !== undefined && (
        <Text className={styles.label} size="md" weight="medium" as="span">
          {label}
        </Text>
      )}
      {/* We use CSS to swap between this navigation hint and the provided
      children on hover - see the styles module. */}
      {(Component === "button" || Component === "a") && (
        <ChevronRightIcon
          width={24}
          height={24}
          className={styles["nav-hint"]}
          aria-hidden={true}
        />
      )}
      {children}
    </Component>
  );
};
