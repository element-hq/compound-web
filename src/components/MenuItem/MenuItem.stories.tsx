/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { Meta, StoryFn } from "@storybook/react";
import ExtensionsIcon from "@vector-im/compound-design-tokens/icons/extensions.svg";
import ChatIcon from "@vector-im/compound-design-tokens/icons/chat.svg";
import SettingsLabel from "@vector-im/compound-design-tokens/icons/settings.svg";

import { MenuItem as MenuItemComponent } from "./MenuItem";
import { Text } from "../Typography/Text";

export default {
  title: "MenuItem",
  component: MenuItemComponent,
  tags: ["autodocs"],
  argTypes: {
    onClick: { action: "clicked!" },
  },
  args: {
    label: "Menu item",
    children: (
      <Text as="span" size="sm">
        99
      </Text>
    ),
    Icon: ChatIcon,
  },
  decorators: [
    (Story: StoryFn) => (
      <div style={{ width: 300 }}>
        <Story />
      </div>
    ),
  ],
} as Meta<typeof MenuItemComponent>;

export const Example = {
  render: () => (
    <div style={{ width: 300 }}>
      <MenuItemComponent Icon={ChatIcon} label="First item">
        <Text as="span" size="sm">
          99
        </Text>
      </MenuItemComponent>
      <MenuItemComponent
        Icon={ExtensionsIcon}
        label="Second item with a name that's quite long"
      />
      <MenuItemComponent Icon={SettingsLabel} label={undefined}>
        <Text as="span" size="sm">
          Third item without a label
        </Text>
      </MenuItemComponent>
    </div>
  ),
};

export const Primary = {
  args: {
    kind: "primary",
  },
};

export const Critical = {
  args: {
    kind: "critical",
  },
};

export const PrimaryDisabled = {
  args: {
    ...Primary.args,
    disabled: true,
  },
};

export const CriticalDisabled = {
  args: {
    ...Critical.args,
    disabled: true,
  },
};

export const WithoutLabel = {
  args: {
    ...Primary.args,
    label: undefined,
  },
};

export const WithALongLabel = {
  args: {
    ...Primary.args,
    label:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
  },
};

export const WithALongLabelAndChildren = {
  args: {
    ...Primary.args,
    label:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, empor incididunt ut labore et dolore magna aliqua.",
    children: (
      <Text as="span" size="sm">
        Longer children too
      </Text>
    ),
  },
};
