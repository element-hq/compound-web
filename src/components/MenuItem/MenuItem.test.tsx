/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import React from "react";
import LeaveIcon from "@vector-im/compound-design-tokens/icons/leave.svg";
import UserProfileIcon from "@vector-im/compound-design-tokens/icons/user-profile.svg";
import MicOnOutlineIcon from "@vector-im/compound-design-tokens/icons/mic-on-outline.svg";

import { MenuItem } from "./MenuItem";
import { Text } from "../Typography/Text";

import * as stories from "./MenuItem.stories";
import { composeStories } from "@storybook/react";

const {
  Primary,
  Critical,
  PrimaryDisabled,
  CriticalDisabled,
  WithALongLabel,
  WithALongLabelAndChildren,
} = composeStories(stories);

describe("MenuItem", () => {
  it("renders", () => {
    const { asFragment } = render(
      <MenuItem kind="critical" Icon={LeaveIcon} label="Leave room" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with a child", () => {
    const { asFragment } = render(
      <MenuItem Icon={UserProfileIcon} label="People">
        <Text as="span" size="sm">
          10
        </Text>
      </MenuItem>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders without a label", () => {
    const { asFragment } = render(
      <MenuItem Icon={MicOnOutlineIcon} label={undefined}>
        Imagine that there might be a volume slider here in place of the label
      </MenuItem>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a Primary menu item", () => {
    const { container } = render(<Primary />);
    expect(container).toMatchSnapshot();
  });
  it("renders a Critical menu item", () => {
    const { container } = render(<Critical />);
    expect(container).toMatchSnapshot();
  });
  it("renders a Primary Disabled menu item", () => {
    const { container } = render(<PrimaryDisabled />);
    expect(container).toMatchSnapshot();
  });
  it("renders a Critical Disabled menu item", () => {
    const { container } = render(<CriticalDisabled />);
    expect(container).toMatchSnapshot();
  });
  it("renders a menu item with a long label", () => {
    const { container } = render(<WithALongLabel />);
    expect(container).toMatchSnapshot();
  });
  it("renders a menu item with a long label and children", () => {
    const { container } = render(<WithALongLabelAndChildren />);
    expect(container).toMatchSnapshot();
  });
});
