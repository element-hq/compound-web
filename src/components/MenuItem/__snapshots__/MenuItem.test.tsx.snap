// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MenuItem > renders 1`] = `
<DocumentFragment>
  <button
    class="_item_37eef5 _interactive_37eef5"
    data-kind="critical"
    role="menuitem"
  >
    <svg
      aria-hidden="true"
      class="_icon_37eef5"
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M14 12.9999C14.2833 12.9999 14.5208 12.904 14.7125 12.7124C14.9042 12.5207 15 12.2832 15 11.9999C15 11.7165 14.9042 11.479 14.7125 11.2874C14.5208 11.0957 14.2833 10.9999 14 10.9999C13.7167 10.9999 13.4792 11.0957 13.2875 11.2874C13.0958 11.479 13 11.7165 13 11.9999C13 12.2832 13.0958 12.5207 13.2875 12.7124C13.4792 12.904 13.7167 12.9999 14 12.9999Z"
        fill="currentColor"
      />
      <path
        d="M10.385 21.7879C10.142 21.5989 10 21.3079 10 20.9999V2.99987C10.0006 2.84816 10.0356 2.69858 10.1024 2.56237C10.1692 2.42615 10.2661 2.30687 10.3856 2.21349C10.5052 2.12011 10.6444 2.05507 10.7927 2.02327C10.941 1.99146 11.0947 1.99372 11.242 2.02987L19.242 4.02987C19.4584 4.08385 19.6506 4.20865 19.788 4.38441C19.9253 4.56017 19.9999 4.77681 20 4.99987V18.9999C20.0001 19.223 19.9255 19.4397 19.7882 19.6155C19.6508 19.7913 19.4585 19.916 19.242 19.9699L11.242 21.9699C11.0947 22.0068 10.941 22.0096 10.7924 21.9781C10.6439 21.9465 10.5046 21.8815 10.385 21.7879ZM18 5.78088L12 4.28088V19.7189L18 18.2189V5.78088Z"
        fill="currentColor"
      />
      <path
        d="M9 5.99988H7L7 17.9999H9V19.9999H7C5.89543 19.9999 5 19.1044 5 17.9999V5.99988C5 4.89531 5.89543 3.99988 7 3.99988H9V5.99988Z"
        fill="currentColor"
      />
    </svg>
    <span
      class="_typography_489030 _font-body-md-medium_489030 _label_37eef5"
    >
      Leave room
    </span>
    <svg
      aria-hidden="true"
      class="_nav-hint_37eef5"
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8.70005 17.3C8.51672 17.1167 8.42505 16.8834 8.42505 16.6C8.42505 16.3167 8.51672 16.0834 8.70005 15.9L12.6 12L8.70005 8.10005C8.51672 7.91672 8.42505 7.68338 8.42505 7.40005C8.42505 7.11672 8.51672 6.88338 8.70005 6.70005C8.88338 6.51672 9.11672 6.42505 9.40005 6.42505C9.68338 6.42505 9.91672 6.51672 10.1 6.70005L14.7 11.3C14.8 11.4 14.8709 11.5084 14.9125 11.625C14.9542 11.7417 14.975 11.8667 14.975 12C14.975 12.1334 14.9542 12.2584 14.9125 12.375C14.8709 12.4917 14.8 12.6 14.7 12.7L10.1 17.3C9.91672 17.4834 9.68338 17.575 9.40005 17.575C9.11672 17.575 8.88338 17.4834 8.70005 17.3Z"
        fill="currentColor"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`MenuItem > renders with a child 1`] = `
<DocumentFragment>
  <button
    class="_item_37eef5 _interactive_37eef5"
    data-kind="primary"
    role="menuitem"
  >
    <svg
      aria-hidden="true"
      class="_icon_37eef5"
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9.175 13.825C9.95833 14.6083 10.9 15 12 15C13.1 15 14.0417 14.6083 14.825 13.825C15.6083 13.0417 16 12.1 16 11C16 9.9 15.6083 8.95833 14.825 8.175C14.0417 7.39167 13.1 7 12 7C10.9 7 9.95833 7.39167 9.175 8.175C8.39167 8.95833 8 9.9 8 11C8 12.1 8.39167 13.0417 9.175 13.825ZM13.4125 12.4125C13.0208 12.8042 12.55 13 12 13C11.45 13 10.9792 12.8042 10.5875 12.4125C10.1958 12.0208 10 11.55 10 11C10 10.45 10.1958 9.97917 10.5875 9.5875C10.9792 9.19583 11.45 9 12 9C12.55 9 13.0208 9.19583 13.4125 9.5875C13.8042 9.97917 14 10.45 14 11C14 11.55 13.8042 12.0208 13.4125 12.4125Z"
        fill="currentColor"
      />
      <path
        d="M22 12C22 17.5228 17.5228 22 12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2C17.5228 2 22 6.47715 22 12ZM20 12C20 7.58172 16.4183 4 12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20C16.4183 20 20 16.4183 20 12Z"
        fill="currentColor"
      />
      <path
        d="M16.2297 18.7917C15.7472 18.6085 15.2623 18.4571 14.775 18.3375C13.8583 18.1125 12.9333 18 12 18C11.0667 18 10.1417 18.1125 9.225 18.3375C8.73774 18.4571 8.25283 18.6085 7.77028 18.7917C7.12696 18.3902 6.54483 17.8998 6.04104 17.3378C6.93176 16.9259 7.83474 16.6092 8.75 16.3875C9.81667 16.1292 10.9 16 12 16C13.1 16 14.1833 16.1292 15.25 16.3875C16.1653 16.6092 17.0682 16.9259 17.959 17.3378C17.4552 17.8998 16.873 18.3902 16.2297 18.7917Z"
        fill="currentColor"
      />
    </svg>
    <span
      class="_typography_489030 _font-body-md-medium_489030 _label_37eef5"
    >
      People
    </span>
    <svg
      aria-hidden="true"
      class="_nav-hint_37eef5"
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8.70005 17.3C8.51672 17.1167 8.42505 16.8834 8.42505 16.6C8.42505 16.3167 8.51672 16.0834 8.70005 15.9L12.6 12L8.70005 8.10005C8.51672 7.91672 8.42505 7.68338 8.42505 7.40005C8.42505 7.11672 8.51672 6.88338 8.70005 6.70005C8.88338 6.51672 9.11672 6.42505 9.40005 6.42505C9.68338 6.42505 9.91672 6.51672 10.1 6.70005L14.7 11.3C14.8 11.4 14.8709 11.5084 14.9125 11.625C14.9542 11.7417 14.975 11.8667 14.975 12C14.975 12.1334 14.9542 12.2584 14.9125 12.375C14.8709 12.4917 14.8 12.6 14.7 12.7L10.1 17.3C9.91672 17.4834 9.68338 17.575 9.40005 17.575C9.11672 17.575 8.88338 17.4834 8.70005 17.3Z"
        fill="currentColor"
      />
    </svg>
    <span
      class="_typography_489030 _font-body-sm-regular_489030"
    >
      10
    </span>
  </button>
</DocumentFragment>
`;

exports[`MenuItem > renders without a label 1`] = `
<DocumentFragment>
  <button
    class="_item_37eef5 _interactive_37eef5 _no-label_37eef5"
    data-kind="primary"
    role="menuitem"
  >
    <svg
      aria-hidden="true"
      class="_icon_37eef5"
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 12C6 11.4477 5.55228 11 5 11C4.44772 11 4 11.4477 4 12C4 16.0796 7.05369 19.446 11 19.9381V21C11 21.5523 11.4477 22 12 22C12.5523 22 13 21.5523 13 21V19.9381C16.9463 19.446 20 16.0796 20 12C20 11.4477 19.5523 11 19 11C18.4477 11 18 11.4477 18 12C18 15.3137 15.3137 18 12 18C8.68629 18 6 15.3137 6 12Z"
        fill="currentColor"
      />
      <path
        clip-rule="evenodd"
        d="M14 12V6C14 4.89543 13.1046 4 12 4C10.8954 4 10 4.89543 10 6V12C10 13.1046 10.8954 14 12 14C13.1046 14 14 13.1046 14 12ZM12 2C9.79086 2 8 3.79086 8 6V12C8 14.2091 9.79086 16 12 16C14.2091 16 16 14.2091 16 12V6C16 3.79086 14.2091 2 12 2Z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
    <svg
      aria-hidden="true"
      class="_nav-hint_37eef5"
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8.70005 17.3C8.51672 17.1167 8.42505 16.8834 8.42505 16.6C8.42505 16.3167 8.51672 16.0834 8.70005 15.9L12.6 12L8.70005 8.10005C8.51672 7.91672 8.42505 7.68338 8.42505 7.40005C8.42505 7.11672 8.51672 6.88338 8.70005 6.70005C8.88338 6.51672 9.11672 6.42505 9.40005 6.42505C9.68338 6.42505 9.91672 6.51672 10.1 6.70005L14.7 11.3C14.8 11.4 14.8709 11.5084 14.9125 11.625C14.9542 11.7417 14.975 11.8667 14.975 12C14.975 12.1334 14.9542 12.2584 14.9125 12.375C14.8709 12.4917 14.8 12.6 14.7 12.7L10.1 17.3C9.91672 17.4834 9.68338 17.575 9.40005 17.575C9.11672 17.575 8.88338 17.4834 8.70005 17.3Z"
        fill="currentColor"
      />
    </svg>
    Imagine that there might be a volume slider here in place of the label
  </button>
</DocumentFragment>
`;
